fsspec/implementations/http.py

```diff
@@ -262,9 +263,19 @@ async def _put_file(
         **kwargs,
     ):
         async def gen_chunks():
-            with open(rpath, "rb") as f:
-                callback.set_size(f.seek(0, 2))
-                f.seek(0)
+            # Support passing arbitrary file-like objects
+            # and use them instead of streams.
+            if isinstance(rpath, io.IOBase):
+                context = nullcontext(rpath)
+                use_seek = False  # might not support seeking
+            else:
+                context = open(rpath, "rb")
+                use_seek = True
```

Comment:
The file instance might happen to have an `.size` attribute - many fsspec implementations do use this.